core/pkg/artifacts/downloader.go

```diff
@@ -86,6 +91,124 @@ func (ad *ArtifactDownloader) getArtifactManifest(artifactID string) (manifest M
 	return manifest, nil
 }
 
+// getBatchEntriesWithFileUrls fetches file urls for the entries in
+// entriesToFetch and returns a list of the formatted manifest entries to
+// download in the current batch.
+func (ad *ArtifactDownloader) getBatchEntriesWithFileUrls(
+	artifactID string,
+	entriesToFetch []gql.ArtifactManifestEntryInput,
+	manifest Manifest,
+	manifestEntries map[string]ManifestEntry,
+	nameToScheduledTime map[string]time.Time,
+) ([]ManifestEntry, error) {
+	var entries []ManifestEntry
+	now := time.Now()
+	response, err := gql.ArtifactFileURLsByManifestEntries(
+		ad.Ctx,
+		ad.GraphqlClient,
+		artifactID,
+		entriesToFetch,
+		manifest.StoragePolicyConfig.StorageLayout,
+		strconv.Itoa(int(manifest.Version)),
+		nil,
+	)
+	if err != nil {
+		return nil, err
+	}
+	for _, edge := range response.GetArtifact().GetFilesByManifestEntries().Edges {
+		filePath := edge.GetNode().Name
+		entry, err := manifest.GetManifestEntryFromArtifactFilePath(filePath)
+		if err != nil {
+			return nil, err
+		}
+		node := edge.GetNode()
+		if node == nil {
+			return nil, fmt.Errorf("error reading entry from fetched file urls")
+		}
+		entry.DownloadURL = &node.DirectUrl
+		entry.LocalPath = &filePath
+		nameToScheduledTime[*entry.LocalPath] = now
+		entries = append(entries, entry)
+		delete(manifestEntries, filePath)
+	}
+	return entries, nil
+}
+
+// collectEntriesForBatch processes BATCH_SIZE manifest entries and returns the gql
+// inputs we need to fetch file urls for and the number of entries skipped. When
+// reference artifacts are handled in core, this should return those entries instead
+// of skipping them.
+func (ad *ArtifactDownloader) collectEntriesForBatch(
+	manifestEntriesCopy map[string]ManifestEntry,
+	nameToScheduledTime map[string]time.Time,
+) ([]gql.ArtifactManifestEntryInput, int) {
```

Comment:
**suggestion (optional, if-minor):** I realize this is only 2 return variables, but if we definitely need to pass `numSkipped` around, what would you think about keeping this in a struct for clarity?  E.g. defining a couple types like:
```go
type batchInputInfo struct {
    Items:      []gql.ArtifactManifestEntryInput
    NumSkipped: int
}

...

type batchInfo struct {
    Items:      []ManifestEntry
    NumSkipped: int
}
```

...then refactoring where needed to pass these structs around, instead of `numSkipped` as a separate variable.

(If you do end up doing this, feel free to change the name ofc, if you wish)